### PR TITLE
Spoi 4443: Using SimpleFilterReusableStreamProvider

### PR DESCRIPTION
**Pull request by DT-Priyanka**

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java [hunk 1]
```diff
@@ -45,6 +51,36 @@ public void finalizeContext() throws IOException
     }
   }
 
+  public static class LZOFilterStreamContext implements FilterStreamContext<LZOpoutputStream>
```
Comment (PramodSSImmaneni) [inline library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java]: Can you derive from BaseFilterStreamContext

File: library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java [hunk 2]
```diff
@@ -45,6 +51,36 @@ public void finalizeContext() throws IOException
...
+    }
+
+    @Override
+    public void finalizeContext() throws IOException
+    {
+      outputStream.flush();
+    }
+  }
+
+  static class LZOpoutputStream extends FilterOutputStream
```
Comment (PramodSSImmaneni) [inline library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java]: Is this being used?

File: library/pom.xml [hunk 3]
```diff
@@ -325,6 +325,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
```
Comment (PramodSSImmaneni) [inline library/pom.xml]: Typically we put components that need dependencies that are not already in dependency tree of dt or hadoop in contrib.

File: library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java [hunk 4]
```diff
@@ -64,6 +100,18 @@ public void reclaimFilterStreamContext(FilterStreamContext<GZIPOutputStream> fil
   }
 
   /**
+   * A provider for LZO filter
```
Comment (PramodSSImmaneni) [inline library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java]: Add documentation for the provider. For example the provider cannot be used if a file is not rewritten from beginning when there is operator failure.

File: library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java [hunk 5]
```diff
@@ -45,6 +51,36 @@ public void finalizeContext() throws IOException
     }
   }
 
+  public static class LZOFilterStreamContext implements FilterStreamContext<LZOpoutputStream>
+  {
+    private LZOpoutputStream outputStream;
+
+    public LZOFilterStreamContext(OutputStream outputStream) throws IOException
+    {
+      this.outputStream = new LZOpoutputStream(new LzopOutputStream(outputStream, LzoLibrary.getInstance().newCompressor(LzoAlgorithm.LZO1X, LzoConstraint.COMPRESSION)));
```
Comment (PramodSSImmaneni) [inline library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java]: Why wrap the Lzo stream into another LZO stream

File: library/pom.xml (hunk 3 above)
Comment (DT-Priyanka) [inline library/pom.xml]: I did not get this, you mean move this dependency to contrib pom file?

File: library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java (hunk 5 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java]: LzopOutputStream extends OutputStream whereas filterStream has to extend from FilterOutputStream, FilterOutputStream is subclass of OutputStream. There was no way I could have initialized filterStream to LzopOutputStream directly. So I did this wrapping.

File: library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java (hunk 2 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/FilterStreamCodec.java]: yes, this is just for object wrapping as explained in other comment.

**Pull request closed without merging**